hips/hip-9999.md

```diff
@@ -0,0 +1,76 @@
+---
+hip: 9999
+title: "Document backwards-compatibility rules"
+authors: [ "Marc Khouzam <marc.khouzam@montreal.ca>" ]
+created: "2020-09-18"
+type: "process"
+status: "draft"
+---
+
+## Abstract
+
+Define and document the backwards-compatibility rules that apply to Helm minor and patch releases.
+
+## Motivation
+
+Helm aims to strictly respect [Semantic Versioning][semver], where minor and patch releases must be 100% backwards-compatible.
+Backwards-compatibility implies that when moving to a new backwards-compatible release:
+* tools (such as Continuous Integration tools) using the helm CLI will not require any changes
+* programs using the Helm Go API will continue to compile 
+
+Some contributors to the Helm code base may not be comfortable with the details of what constitutes a compatibility-breaking change.
+Furthermore, in some cases, there can be a difference of opinion, even amongst the maintainers, on what constitutes such a change.
+
+Having a well-defined, documented set of backwards-compatibility rules for the Helm code-base will allow to:
+* avoid compatibility-breaking changes being merged by mistake
+* guide contributors in understanding helm's backwards-compatibility requirements
+* remove any uncertainty when it comes to enforcing backwards compatibility
+* provide a basis for discussion and possible updates to how helm handles backwards compatibility
+
+## Rationale
+
+The maintainers have agreed that a formal definition of Helm's backwards-compatibility rules would be beneficial.
+
+## Specification
+
+The following compatibility rules will apply to helm minor and patch releases:
+* exported Go APIs will remain compatible as per this [article][go-module-comp]
+* CLI commands:
+    * are case-sensitive
+    * must not be removed
+    * must not be renamed
+    * must not be moved elsewhere in the command hierarchy
+* CLI flags:
```

Comment:
Also for this list:

- must have a long name (e.g. `--foo`, not just `-f`)
- should not have a short name unless the flag is frequently used (thus, `install` only has 3 with single-letter abbreviations (this is a reason to reject a change PR)
- must not be added/moved to the global context unless they impact _all_ Helm commands (with exceptions for global flags that were added before this policy)

Reply:
> * must have a long name (e.g. --foo, not just -f)
> * should not have a short name unless the flag is frequently used (thus, install only has 3 with single-letter abbreviations (this is a reason to reject a change PR)
> * must not be added/moved to the global context unless they impact all Helm commands (with exceptions for global flags that were added before this policy)

All valid points, but I think they go beyond backwards compatibility and fall into a design rules category.

But I've added to the backwards-compatibility rules that the flag section applies as much to long-name flags as short-name flags, and that one form cannot be changed even if the other stays the same

Reply:
These are all valid points.  However, they all seem to be global design rules, not necessarily limited to backwards compatibility.  We have to decide where to put such design rules.